spec/app/ics-721-nft-transfer/README.md

```diff
@@ -0,0 +1,386 @@
+- When acting as the sink zone, the bridge module burns the local voucher on the sending chain and unescrows the local non-fungible token on the receiving chain.
+- When a packet times-out, local non-fungible tokens are unescrowed back to the sender or vouchers minted back to the sender appropriately.
+- Acknowledgement data is used to handle failures, such as invalid destination accounts. Returning
+  an acknowledgement of failure is preferable to aborting the transaction since it more easily enables the sending chain to take appropriate action based on the nature of the failure.
+
+`createOutgoingPacket` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
+
+```typescript
+function createOutgoingPacket(
+  classId: string,
+  tokenIds: []string,
+  sender: string,
+  receiver: string,
+  source: boolean,
+  destPort: string,
+  destChannel: string,
+  sourcePort: string,
+  sourceChannel: string,
+  timeoutHeight: Height,
+  timeoutTimestamp: uint64) {
+  prefix = "{sourcePort}/{sourceChannel}/"
+  // we are the source if the classId is not prefixed
+  source = classId.slice(0, len(prefix)) !== prefix
+  tokenUris = []
+  for (let tokenId in tokenIds) {
+    // assert that sender is token owner
+    abortTransactionUnless(sender === nft.getOwner(classId, tokenId))
+    if source {
+      // determine escrow account
+      escrowAccount = channelEscrowAddresses[sourceChannel]
+      // escrow source token
+      nft.Transfer(classId, tokenId, escrowAccount)
+    } else {
+      // receiver is source chain, burn voucher
+      bank.Burn(classId, tokenId)
+    }
+    tokenUris.push(nft.getNFT(classId, tokenId).getUri())
+  }
+  NonFungibleTokenPacketData data = NonFungibleTokenPacketData{classId, nft.getClass(classId).getUri(), tokenIds, tokenUris, sender, receiver}
+  handler.sendPacket(Packet{timeoutHeight, timeoutTimestamp, destPort, destChannel, sourcePort, sourceChannel, data}, getCapability("port"))
+}
+```
+
+`onRecvPacket` is called by the routing module when a packet addressed to this module has been received.
+
+```typescript
+function onRecvPacket(packet: Packet) {
+  NonFungibleTokenPacketData data = packet.data
+  // construct default acknowledgement of success
+  NonFungibleTokenPacketAcknowledgement ack = NonFungibleTokenPacketAcknowledgement{true, null}
+  prefix = "{packet.sourcePort}/{packet.sourceChannel}/"
+  // we are the source if the packets were prefixed by the sending chain
+  source = data.classId.slice(0, len(prefix)) === prefix
+  for (var i in data.tokenIds) {
+    if source {
+      // receiver is source chain: unescrow token
+      // determine escrow account
+      escrowAccount = channelEscrowAddresses[packet.destChannel]
+      // assert that escrow account is token owner
+      abortTransactionUnless(escrowAccount === nft.getOwner(data.classId.slice(len(prefix)), data.tokenIds[i]))
```

Comment:
We should remove this line so that if transfer fails because module account is not owner, we will write failed acknowledgement rather than just abort tx